rbc/rbc.go

```diff
@@ -1,100 +1,483 @@
 package rbc
 
 import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+
 	"github.com/DE-labtory/cleisthenes"
 	"github.com/DE-labtory/cleisthenes/pb"
+	"github.com/DE-labtory/cleisthenes/rbc/merkletree"
+	"github.com/DE-labtory/iLogger"
+	"github.com/golang/protobuf/ptypes"
 	"github.com/klauspost/reedsolomon"
+	"github.com/syndtr/goleveldb/leveldb/errors"
 )
 
-type RBC struct {
+var ErrInvalidRBCType = errors.New("Invalid RBC message type")
 
+type RBC struct {
 	// number of network nodes
 	n int
 
 	// number of byzantine nodes which can tolerate
 	f int
 
+	// owner of rbc instance (node)
+	owner cleisthenes.Member
+
+	// proposerId is the ID of proposing node
 	proposer cleisthenes.Member
 
 	// Erasure coding using reed-solomon method
 	enc reedsolomon.Encoder
 
-	// Broadcast message after when rbc receive message
-	messages []*pb.Message
+	// output of RBC
+	value []byte
+
+	// length of original data
+	contentLength uint64
+
+	// number of sharded data and parity
+	// data : N - F, parity : F
+	numDataShards, numParityShards int
 
 	// Request of other rbcs
-	valueReqRepo cleisthenes.RequestRepository
 	echoReqRepo  cleisthenes.RequestRepository
 	readyReqRepo cleisthenes.RequestRepository
 
+	valReceived, echoSent, readySent bool
+
 	// internal channels to communicate with other components
-	closeChan   chan struct{}
-	messageChan chan InnerMessage
-	requestChan chan InnerRequest
+	closeChan chan struct{}
+	reqChan   chan request
+	inputChan chan InputMessage
+
+	broadcaster cleisthenes.Broadcaster
+}
+
+func New(n, f int, owner, proposer cleisthenes.Member, broadcaster cleisthenes.Broadcaster) *RBC {
+	numParityShards := f
+	numDataShards := n - numParityShards
+
+	enc, err := reedsolomon.New(numDataShards, numParityShards)
+	if err != nil {
+		panic(err)
+	}
+
+	echoReqRepo, err := NewEchoReqRepository()
+	if err != nil {
+		panic(err)
+	}
+
+	readyReqRepo, err := NewReadyReqRepository()
+	if err != nil {
+		panic(err)
+	}
+
+	rbc := &RBC{
+		n:               n,
+		f:               f,
+		owner:           owner,
+		proposer:        proposer,
+		enc:             enc,
+		numDataShards:   numDataShards,
+		numParityShards: numParityShards,
+		echoReqRepo:     echoReqRepo,
+		readyReqRepo:    readyReqRepo,
+		closeChan:       make(chan struct{}),
+		reqChan:         make(chan request),
+		inputChan:       make(chan InputMessage),
+		broadcaster:     broadcaster,
+	}
+	go rbc.run()
+	return rbc
+}
+
+func (rbc *RBC) distributeMessage(proposer cleisthenes.Member, reqs []cleisthenes.Request) error {
+	if len(reqs) != rbc.n {
+		return fmt.Errorf("invalid message count - expected : %d, got : %d", rbc.n, len(reqs))
+	}
+
+	var typ pb.RBCType
+	switch reqs[0].(type) {
+	case *ValRequest:
+		typ = pb.RBC_VAL
+	default:
+		return errors.New("invalid distributeMessage message type")
+	}
+
+	if err := rbc.handleValueRequest(rbc.proposer, reqs[0].(*ValRequest)); err != nil {
```

Comment:
why handling ValueRequest here??

Reply:
if network has `n` nodes, In rbc's `distributeMessage`, broadcast each `n` sharded request to each nodes. and `n` sharded request include own self. so `handleValueRequest` send request to itself. 

and i think rbc's `distributeMessage` has to handle above thing - handle own self in `distributeMessage`

Reply:
1. `distributeMessage` should only cares about distributing message not handling value request.

2. I think sending value request from myself should move on to `HandleInput` 

```
reqChan <- request{
		sender: rbc.owner,
		data:   req,
		err:    make(chan error),
	}

distributeMessage()
``` 

because function name does not contain the context handling value message from myself

Reply:
1. you mean `pb.message`?

2. i agree